src/Illuminate/Database/Eloquent/Relations/HasOneOrManyThrough.php

```diff
@@ -97,12 +97,16 @@ public function __construct(Builder $query, Model $farParent, Model $throughPare
      */
     public function addConstraints()
     {
+        $query = $this->getRelationQuery();
         $localValue = $this->farParent[$this->localKey];
 
-        $this->performJoin();
+        // Join is performed in HasOneThrough::addOneOfManySubQueryConstraints
+        if (! method_exists($this, 'isOneOfMany') || ! $this->isOneOfMany()) {
+            $this->performJoin($query);
+        }
```

Comment:
This is admittedly not very pretty, but is necessary unless the `Relation::addConstraints` signature is changed to allow passing the query instance in (see Additional Considerations section in PR description).

The root problem is that **all** subqueries need the join applied, but this method is only called for the base query and the innermost subquery.